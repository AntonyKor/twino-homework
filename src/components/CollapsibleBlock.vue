<template>
  <div :class="{ collapsed: isCollapsed }" class="collapsible">
    <header>
      <h3>{{ title }}</h3>
      <button @click="isCollapsed = !isCollapsed">
        Show {{ isCollapsed ? "more" : "less" }}
      </button>
    </header>
    <div class="content"><slot></slot></div>
  </div>
</template>

<script>
export default {
  props: {
    title: {
      type: String,
      default: "",
    },
  },

  data() {
    return { isCollapsed: true };
  },
};
</script>

<style scoped lang="scss">
.collapsible {
  padding: 30px 0;
  border-bottom: 1px solid #9e9fa2;
}

header {
  display: flex;
  justify-content: space-between;
  align-items: center;

  button {
    margin-left: 25px;
    position: relative;
    background: none;
    border: none;
    outline: none;
    color: #2e50b6;
    font-weight: bold;
    cursor: pointer;

    &:before {
      position: absolute;
      content: "";
      display: block;
      width: 10px;
      height: 3px;
      background-color: #2e50b6;
      top: 50%;
      left: -15px;
      transform: translateY(-50%);
    }
  }
}

.content {
  margin-top: 30px;
}

.collapsed {
  .content {
    height: 0px;
    margin-top: 0;
    overflow: hidden;
  }

  header button:after {
    position: absolute;
    content: "";
    display: block;
    width: 10px;
    height: 3px;
    background-color: #2e50b6;
    top: 50%;
    left: -15px;
    transform: translateY(-50%) rotate(0.25turn);
  }
}
</style>
