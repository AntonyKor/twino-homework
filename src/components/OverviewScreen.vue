<template>
  <div>
    <h2>Answers owerview</h2>
    <CollapsibleBlock
      v-for="(group, groupIndex) in groups"
      :key="group.title"
      :title="group.title"
    >
      <div
        v-for="(question, questionIndex) in group.questions"
        :key="question.question"
      >
        <h4>{{ question.question }}</h4>
        <p>
          <a
            @click="$emit('question-select', { groupIndex, questionIndex })"
            href="#"
            >{{ question.answers[question.answer] }}</a
          >
        </p>
      </div>
    </CollapsibleBlock>
  </div>
</template>

<script>
import CollapsibleBlock from "./CollapsibleBlock.vue";

export default {
  components: {
    CollapsibleBlock,
  },
  props: {
    groups: {
      type: Array,
      required: true,
    },
  },
};
</script>

<style scoped lang="scss">
h2 {
  font-size: 2rem;
}

h4 {
  margin: 30px 0 10px;
}

p {
  margin: 0;
}

a {
  color: #9e9fa2;
  text-decoration: none;
}
</style>
