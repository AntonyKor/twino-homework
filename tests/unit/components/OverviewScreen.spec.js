import { shallowMount } from "@vue/test-utils";
import OverviewScreen from "@/components/OverviewScreen.vue";
import CollapsibleBlock from "@/components/CollapsibleBlock.vue";

describe("OverviewScreen", () => {
  const groups = [
    {
      title: "group1",
      questions: [
        { question: "question1", answers: ["a1", "a2"], answer: 1 },
        { question: "question2", answers: ["a3"], answer: 0 },
      ],
    },
    {
      title: "group2",
      questions: [{ question: "question3", answers: ["a5", "a6"], answer: 0 }],
    },
  ];

  it("should match snapshot", () => {
    const wrapper = shallowMount(OverviewScreen, {
      propsData: { groups },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render CollapsibleBlock for every group", () => {
    const wrapper = shallowMount(OverviewScreen, {
      propsData: { groups },
    });

    expect.hasAssertions();
    const collapsibleBlocks = wrapper.findAllComponents(CollapsibleBlock);
    groups.forEach((group, i) =>
      expect(collapsibleBlocks.at(i).props("title")).toEqual(group.title)
    );
  });

  it("should render every question with answer", () => {
    const wrapper = shallowMount(OverviewScreen, {
      propsData: { groups },
    });

    expect.hasAssertions();
    const collapsibleBlocks = wrapper.findAllComponents(CollapsibleBlock);
    groups.forEach((group, i) =>
      group.questions.forEach((question, j) => {
        expect(collapsibleBlocks.at(i).findAll("h4").at(j).text()).toEqual(
          question.question
        );
        expect(collapsibleBlocks.at(i).findAll("a").at(j).text()).toEqual(
          question.answers[question.answer]
        );
      })
    );
  });

  it("should emit 'question-select' event on click on a", async () => {
    const wrapper = shallowMount(OverviewScreen, {
      propsData: { groups },
    });

    await wrapper
      .findAllComponents(CollapsibleBlock)
      .at(1)
      .find("a")
      .trigger("click");

    expect(wrapper.emitted("question-select")).toHaveLength(1);
    expect(wrapper.emitted("question-select")[0]).toEqual([
      {
        groupIndex: 1,
        questionIndex: 0,
      },
    ]);
  });
});
