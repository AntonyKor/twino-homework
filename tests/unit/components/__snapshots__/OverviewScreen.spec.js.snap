// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewScreen should match snapshot 1`] = `
<div>
  <h2>
    Answers owerview
  </h2>
   
  <collapsibleblock-stub
    title="group1"
  >
    <div>
      <h4>
        question1
      </h4>
       
      <p>
        <a
          href="#"
        >
          a2
        </a>
      </p>
    </div>
    <div>
      <h4>
        question2
      </h4>
       
      <p>
        <a
          href="#"
        >
          a3
        </a>
      </p>
    </div>
  </collapsibleblock-stub>
  <collapsibleblock-stub
    title="group2"
  >
    <div>
      <h4>
        question3
      </h4>
       
      <p>
        <a
          href="#"
        >
          a5
        </a>
      </p>
    </div>
  </collapsibleblock-stub>
</div>
`;
